src/hb-kbts.cc

```diff
@@ -0,0 +1,248 @@
+  if (unlikely (!blob_length))
+  {
+    DEBUG_MSG (KBTS, blob, "Empty blob");
+    goto fail;
+  }
+
+  data = hb_malloc (blob_length);
+  if (unlikely (!data))
+    goto fail;
+
+  kb_font = (kbts_font *) hb_malloc (sizeof(kbts_font));
+  if (unlikely (!kb_font))
+  {
+    hb_free (data);
+    goto fail;
+  }
+
+  // Copy the blob data because kbts_ReadFontData() will modify it.
+  hb_memcpy (data, blob_data, blob_length);
+
+  // The `data` and `memory` pointers will be freed by kbts_FreeFont().
+
+  scratch_size = kbts_ReadFontHeader (kb_font, data, blob_length);
+  scratch = hb_malloc (scratch_size);
+  memory_size = kbts_ReadFontData (kb_font, scratch, scratch_size);
+
+  memory = scratch;
+  if (memory_size > scratch_size)
+    memory = hb_realloc (memory, memory_size);
+
+  kbts_PostReadFontInitialize (kb_font, memory, memory_size);
+
+  if (unlikely (!kbts_FontIsValid(kb_font)))
+  {
+    DEBUG_MSG (KBTS, face, "kbts_font creation failed");
+    _hb_kbts_shaper_face_data_destroy ((hb_kbts_face_data_t *) kb_font);
+    kb_font = nullptr;
+    goto fail;
+  }
+
+fail:
+  hb_blob_destroy (blob);
+
+  return (hb_kbts_face_data_t *) kb_font;
+}
+
+void
+_hb_kbts_shaper_face_data_destroy (hb_kbts_face_data_t *data)
+{
+  kbts_font *font = (kbts_font *) data;
+  hb_free (font->FileBase);
+  hb_free (font->GlyphLookupMatrix);
+}
+
+hb_kbts_font_data_t *
+_hb_kbts_shaper_font_data_create (hb_font_t *font)
+{
+  return (hb_kbts_font_data_t *) HB_SHAPER_DATA_SUCCEEDED;
+}
+
+void
+_hb_kbts_shaper_font_data_destroy (hb_kbts_font_data_t *data)
+{
+}
+
+hb_bool_t
+_hb_kbts_shape (hb_shape_plan_t    *shape_plan,
+		hb_font_t          *font,
+		hb_buffer_t        *buffer,
+		const hb_feature_t *features,
+		unsigned int        num_features)
+{
+  hb_face_t *face = font->face;
+  kbts_font *kb_font = (kbts_font *) (const void *) face->data.kbts;
+  if (unlikely (!kb_font))
+  {
+    DEBUG_MSG (KBTS, face, "No kbts font data");
+    return false;
+  }
+
+  kbts_direction kb_direction;
+  switch (buffer->props.direction)
+  {
+    case HB_DIRECTION_LTR: kb_direction = KBTS_DIRECTION_LTR; break;
+    case HB_DIRECTION_RTL: kb_direction = KBTS_DIRECTION_RTL; break;
+    case HB_DIRECTION_TTB:
+    case HB_DIRECTION_BTT:
+      DEBUG_MSG (KBTS, face, "Vertical direction is not supported");
+      return false;
+    case HB_DIRECTION_INVALID:
+      DEBUG_MSG (KBTS, face, "Invalid direction");
+      return false;
+    default: break;
+  }
+
+  kbts_script kb_script = KBTS_SCRIPT_DONT_KNOW;
+  kbts_language kb_language = KBTS_LANGUAGE_DEFAULT;
+  {
+    hb_tag_t script;
+    hb_tag_t language;
+    unsigned int script_count = 1;
+    unsigned int language_count = 1;
+    hb_ot_tags_from_script_and_language (buffer->props.script, buffer->props.language,
+					 &script_count, &script,
+					 &language_count, &language);
+    if (script_count)
+      kb_script = kbts_ScriptTagToScript (hb_uint32_swap (script));
```

Comment:
Does the swap need to be there unconditionally? Or only on LE systems? I forget.